### PR TITLE
Generalize filter strategies

### PR DESCRIPTION
**Pull request by aklish**
This is the first of several restructurings for more general filtering

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java [hunk 1]
```diff
@@ -207,4 +221,17 @@ default void preCommit() {
 
         return val;
     }
+
+    default <T> Object getRelationWithSortingAndPagination(
```
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: I think it would be nice not to require a separate method for handling sorting and pagination. I think this entered the code as more of a band-aid. If we're going to go through a major refactoring of this in the first place, then we should consider consolidating this functionality.

One effective way may be to always give sane defaults in the event sorting and/or pagination is not specified.

Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: Sounds good - will fix

File: elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/DefaultFilterStrategy.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.core.filter.strategy;
+
+import com.yahoo.elide.core.EntityDictionary;
+import com.yahoo.elide.core.filter.Operator;
+import com.yahoo.elide.core.filter.Predicate;
+import com.yahoo.elide.core.filter.expression.AndExpression;
+import com.yahoo.elide.core.filter.expression.Expression;
```
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/DefaultFilterStrategy.java]: `Expression` feels to generic. As I read through the code, I keep finding myself having to recall what namespace this `Expression` is under. Should we rename this to something like `FilterExpression`? Similarly, for Elide 3.0 we should consider this update for Permission `Expression`'s as well.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/JoinFilterStrategy.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * both predicates.
+ *
+ * This filter strategy is invoked whenever the first entity is loaded from the DataStoreTransaction.
+ * For example, the above filter on '/books' would be invoked when 'book' is loaded from the DataStoreTransaction.
+ */
+public interface JoinFilterStrategy {
+    /**
+     * @param path the URL path
+     * @param queryParams the subset of query parameters that start with 'filter'
+     * @return
```
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/JoinFilterStrategy.java]: Missing description of what should be returned. Since this is an interface, we should make sure this is well articulated.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/SubqueryFilterStrategy.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * This strategy is invoked whenever a collection of elements is returned via Elide.  This includes:
+ *  - GET on a collection whose type is referenced in the filter
+ *  - GET on a relationship whose type is referenced in the filter
+ *  - A compound document includes a relationship whose type is referenced in the filter
+ */
+public interface SubqueryFilterStrategy {
+    /**
+     * @param path The URL path
+     * @param queryParams The subset of queryParams that start with 'filter'
+     * @return
```
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/SubqueryFilterStrategy.java]: Same comment as above. This should be well-defined for interface.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 5]
```diff
@@ -346,6 +365,80 @@ public T getObject() {
...
     @Override
+    public <T> Object getRelation(
+            Object entity,
+            RelationshipType relationshipType,
+            String relationName,
+            Class<T> relationClass,
+            EntityDictionary dictionary,
+            Optional<Expression> filterExpression
+    ) {
+        Object val = com.yahoo.elide.core.PersistentResource.getValue(entity, relationName, dictionary);
```
Comment (DeathByTape) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Doesn't this method get called from `PersistentResource`? Would `getValue` be more appropriately placed in the `EntityDictionary` (assuming it's not "checked")?

Realistically, `getValue` is just reflection code that extracts the value from a POJO. We should probably isolate the notion of looking up values in beans and the security around being able to access that value anyway.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 6]
```diff
@@ -346,6 +365,80 @@ public T getObject() {
...
+            Object entity,
+            RelationshipType relationshipType,
+            String relationName,
+            Class<T> relationClass,
+            EntityDictionary dictionary,
+            Optional<Expression> filterExpression,
+            Sorting sorting,
+            Pagination pagination
+    ) {
+        Object val = com.yahoo.elide.core.PersistentResource.getValue(entity, relationName, dictionary);
```
Comment (DeathByTape) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Same comment as above.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java (hunk 5 above)
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: I agree but I think this is out of scope for this PR.  I don't want to boil the ocean (PR is already pretty massive).

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java (hunk 6 above)
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Agree but maybe in a different PR.

**Comment by DeathByTape** [comment]
Since those other changes are going to be postponed to a later request, LGTM.

👍

File: elide-core/src/main/java/com/yahoo/elide/core/RequestScope.java [hunk 7]
```diff
@@ -94,13 +114,36 @@ public RequestScope(JsonApiDocument jsonApiDocument,
...
+            /* Extract any query param that starts with 'filter' */
+            MultivaluedMap filterParams = getFilterParams(queryParams);
+
+            if (! filterParams.isEmpty()) {
+
+                /* First check to see if there is a global, cross-type filter */
+                try {
+                    globalFilterExpression = filterStrategy.parseGlobalExpression(path, filterParams);
+                } catch (ParseException e) {
+                    //TODO - perhaps aggregate this error message with any subsequent filter parsing error */
```
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/RequestScope.java]: We should probably honor this TODO right here. It could make debugging difficult if we only ever log the failure for a single strategy.

**Comment by DeathByTape** [comment]
👍  LGTM but it's pretty large. 

Would be worth getting a second set of reviewer eyes on this from either @clayreimann or @wcekan

File: elide-core/src/main/java/com/yahoo/elide/Elide.java [hunk 8]
```diff
@@ -119,13 +138,19 @@ protected Elide(AuditLogger auditLogger,
                   DataStore dataStore,
                   EntityDictionary dictionary,
                   JsonApiMapper mapper,
-                  Function<RequestScope, PermissionExecutor> permissionExecutor) {
+                  Function<RequestScope, PermissionExecutor> permissionExecutor,
+                  List<JoinFilterStrategy> joinFilterStrategies,
+                  List<SubqueryFilterStrategy> subqueryFilterStrategies,
+                  boolean useFilterExpressions) {
```
Comment (clayreimann) [inline elide-core/src/main/java/com/yahoo/elide/Elide.java]: This is a breaking change because sub-classes could be depending on this constructor.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/MultipleFilterStrategy.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ */
+@AllArgsConstructor
+@Slf4j
+public class MultipleFilterStrategy implements JoinFilterStrategy, SubqueryFilterStrategy {
+    private List<JoinFilterStrategy> joinStrategies;
+    private List<SubqueryFilterStrategy> subqueryStrategies;
+
+    public MultipleFilterStrategy(EntityDictionary dictionary) {
+        DefaultFilterStrategy defaultStrategy = new DefaultFilterStrategy(dictionary);
+        joinStrategies = new ArrayList<>();
```
Comment (clayreimann) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/MultipleFilterStrategy.java]: Is there a reason not to inline these?

File: elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/HQLFilterOperation.java [hunk 10]
```diff
@@ -60,4 +65,47 @@ public String applyAll(Set<Predicate> predicates) {
...
+    public String apply(FilterExpression filterExpression) {
+        HQLQueryVisitor visitor = new HQLQueryVisitor();
+        return "WHERE " + filterExpression.accept(visitor);
+
+    }
+
+    /**
+     * Filter expression visitor which builds an HQL query.
+     */
+    public class HQLQueryVisitor implements Visitor<String> {
```
Comment (clayreimann) [inline elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/HQLFilterOperation.java]: `public static class`?

Comment (aklish) [inline elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/HQLFilterOperation.java]: THis visitor calls apply on the HQLFilterOperation - so it cannot be static as it stands.

File: elide-core/src/main/java/com/yahoo/elide/Elide.java (hunk 8 above)
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/Elide.java]: Fixed

File: elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/MultipleFilterStrategy.java (hunk 9 above)
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/strategy/MultipleFilterStrategy.java]: I usually only inline if I have not defined a constructor.

**Comment by wcekan** [comment]
@aklish Looks like a good direction to take.  Will we add some examples and more IT tests?

**Comment by aklish** [comment]
Existing it tests cover the code.  When I add FIQL, I'll add more.

File: elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/HQLFilterOperation.java (hunk 10 above)
Comment (clayreimann) [inline elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/HQLFilterOperation.java]: Ahh, missed that

**Comment by DeathByTape** [comment]
👍

**Comment by wcekan** [comment]
👍  Ok
